src/treewalk.coffee

```diff
@@ -197,35 +217,101 @@ exports.createTreewalkParser = (parse, config, root) ->
 
         for child in node.children
           @mark child, prefix, depth + 2, false
+
       else if context? and @detNode(context) is 'block'
         if @det(node) is 'socket' and ((not config.SHOULD_SOCKET?) or config.SHOULD_SOCKET(@opts, node))
           @addSocket
+            empty: config.EMPTY_STRINGS?[node.type] ? config.empty
             bounds: node.bounds
             depth: depth
             parseContext: rules[0]
 
-          if config.empty? and not @opts.preserveEmpty and helper.clipLines(@lines, node.bounds.start, node.bounds.end) is config.empty
+          if config.EMPTY_STRINGS? and not @opts.preserveEmpty and helper.clipLines(@lines, node.bounds.start, node.bounds.end) is config.empty
             @flagToRemove node.bounds, depth + 1
 
   if config.droppabilityGraph?
+    # DroppabilityGraph contains all the rules for what things can be
+    # other things (strictly) in the grammar. For instance a * b is a multiplicativeExpression in C,
+    # but can also play the role of an additiveExpression, or an expression. In this case,
+    # the graph would contain edges pointing from expression to additiveExpression, and additiveExpression to multiplicativeExpression.
+    #
+    # This allows us to do a graph search to determine whether A can, transitively, be dropped in B.
     droppabilityGraph = new Graph(config.droppabilityGraph)
+
+    # parenGraph is DroppabilityGraph with paren rules added.
+    #
+    # Paren edges point from things like primaryExpression -> expression, indicating that you can
+    # from a primaryExpression from an expression by adding parentheses. In this case, it would be adding '(' and ')'.
+    #
+    # When we do paren wrapping, we walk through the graph and add any necessary parentheses to get from our AST context
+    # to the desintation.
+    parenGraph = new Graph(config.parenGraph)
+
     TreewalkParser.drop = (block, context, pred) ->
+      if block.parseContext is '__comment__' and context.type in ['indent', 'document']
+        return helper.ENCOURAGE
+      else if context.parseContext is '__comment__'
+        return helper.DISCOURAGE
+
       parseContext = context.indentContext ? context.parseContext
-      if helper.dfs(droppabilityGraph, parseContext, block.parseContext)
+      if helper.dfs(parenGraph, parseContext, block.nodeContext.type)
         return helper.ENCOURAGE
       else
         return helper.FORBID
 
+    TreewalkParser.parens = (leading, trailing, node, context) ->
```

Comment:
Maybe we should rename `context` -> `parent`? One of the things having the argument gives us it that `parens` can be called to "prepare" a node for dropping into a new spot before it actually gets dropped there (so `node.parent != context`).